ts-client/src/amm/index.ts

```diff
@@ -0,0 +1,918 @@
+      .postInstructions(postInstructions)
+      .transaction();
+
+    return new Transaction({
+      feePayer: owner,
+      ...(await this.program.provider.connection.getLatestBlockhash()),
+    }).add(depositTx);
+  }
+
+  /**
+   * `getWithdrawQuote` is a function that takes in a withdraw amount and returns the amount of tokens
+   * that will be withdrawn from the pool
+   * @param {BN} withdrawTokenAmount - The amount of tokens you want to withdraw from the pool.
+   * @param {PublicKey} [tokenMint] - The token you want to withdraw. If you want balanced withdraw, leave this blank.
+   * @param {number} [slippage] - The amount of slippage you're willing to accept. (Max to 2 decimal place)
+   * @returns The return value is a tuple of the poolTokenAmountIn, tokenAOutAmount, and
+   * tokenBOutAmount.
+   */
+  public getWithdrawQuote(withdrawTokenAmount: BN, slippage: number, tokenMint?: PublicKey): WithdrawQuote {
+    const vaultAWithdrawableAmount = calculateWithdrawableAmount(this.accountsInfo.currentTime, this.vaultA.vaultState);
+    const vaultBWithdrawableAmount = calculateWithdrawableAmount(this.accountsInfo.currentTime, this.vaultB.vaultState);
+
+    // balance withdraw
+    if (!tokenMint) {
+      const vaultALpBurn = this.getShareByAmount(
+        withdrawTokenAmount,
+        this.accountsInfo.poolLpSupply,
+        this.accountsInfo.poolVaultALp,
+      );
+      const vaultBLpBurn = this.getShareByAmount(
+        withdrawTokenAmount,
+        this.accountsInfo.poolLpSupply,
+        this.accountsInfo.poolVaultBLp,
+      );
+
+      const tokenAOutAmount = this.getAmountByShare(
+        vaultALpBurn,
+        vaultAWithdrawableAmount,
+        this.accountsInfo.vaultALpSupply,
+      );
+      const tokenBOutAmount = this.getAmountByShare(
+        vaultBLpBurn,
+        vaultBWithdrawableAmount,
+        this.accountsInfo.vaultBLpSupply,
+      );
+
+      return {
+        poolTokenAmountIn: withdrawTokenAmount,
+        tokenAOutAmount: getMinAmountWithSlippage(tokenAOutAmount, slippage),
+        tokenBOutAmount: getMinAmountWithSlippage(tokenBOutAmount, slippage),
+      };
+    }
+
+    // Imbalance withdraw
+    const isWithdrawingTokenA = tokenMint.equals(new PublicKey(this.tokenA.address));
+    const isWithdrawingTokenB = tokenMint.equals(new PublicKey(this.tokenB.address));
+    invariant(isWithdrawingTokenA || isWithdrawingTokenB, ERROR.INVALID_MINT);
+
+    const tradeDirection = tokenMint.equals(this.poolState.tokenAMint) ? TradeDirection.BToA : TradeDirection.AToB;
+
+    const outAmount = this.swapCurve.computeWithdrawOne(
+      withdrawTokenAmount,
+      this.accountsInfo.poolLpSupply,
+      this.poolInfo.tokenAAmount,
+      this.poolInfo.tokenBAmount,
+      this.poolState.fees,
+      tradeDirection,
+    );
+
+    const [vaultLpSupply, vaultTotalAmount] =
+      tradeDirection == TradeDirection.AToB
+        ? [this.accountsInfo.vaultALpSupply, vaultBWithdrawableAmount]
```

Comment:
I think I made a mistake here. It is `[this.accountsInfo.vaultBLpSupply, vaultBWithdrawableAmount]` and `[this.accountsInfo.vaultALpSupply, vaultAWithdrawableAmount];`